azure-iot-hub/tests/test_digital_twin_client.py

```diff
@@ -107,6 +112,63 @@ def test_instantiates_with_connection_string_no_shared_access_key(self):
             DigitalTwinClient(connection_string)
 
 
+@pytest.mark.describe("DigitalTwinClient - .from_connection_string()")
+class TestFromConnectionString:
+    @pytest.mark.it(
+        "Creates an instance of ConnectionStringAuthentication and passes it to IotHubGatewayServiceAPIs constructor"
+    )
+    def test_connection_string_auth(self, mocker):
+        connection_string_auth_init_mock = mocker.patch.object(
+            azure.iot.hub.auth, "ConnectionStringAuthentication"
+        )
+        connection_string_auth_mock = connection_string_auth_init_mock.return_value
+        connection_string_auth_mock.__getitem__.return_value = fake_hostname
+        protocol_client_init_mock = mocker.patch.object(
```

Comment:
I see this mock is probably why you changed the import in the implementation - but is this necessary? I see it's helpful for allowing you to check the params of the instantiation, but could you not just check the value that gets set on the IotHubGatewayServiceAPIs object without mocking?

In general, only mock when necessary. Mock things that are sufficiently complex you can't (or don't want to) instantiate them. Mock things that are strictly necessary for you to demonstrate the requirement in the `pytest.mark.it` statement. Otherwise, mocking is best to be avoided lest you make things more complicated than they need to be.

(I could be wrong about this specific case, maybe it is necessary, I'm just speaking in general here)